### PR TITLE
feat(query): copy into add option max_files(list max files)

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

support:

copy into xxx from s3://file max_files=10;

max_files default is 0. if 0 means list all files.

Closes #issue

### CODE REVIEW

File: src/common/storage/src/stage.rs [hunk 1]
```diff
@@ -107,9 +107,15 @@ impl StageFilesInfo {
         }
     }
 
-    pub async fn list(&self, operator: &Operator, first_only: bool) -> Result<Vec<StageFileInfo>> {
+    pub async fn list(
+        &self,
+        operator: &Operator,
+        first_only: bool,
+        file_limit: usize,
```
Comment (youngsofun) [inline src/common/storage/src/stage.rs:114]: should not limit here,  some may already be copied.
limit should be applied when filtering colored files.

**Comment by youngsofun** [comment]
what about naming it `max_num_files ` or `max_files`?

**Comment by TCeason** [comment]
> max_files

How about max_stages ?

**Comment by BohuTANG** [comment]
> > max_files
> 
> How about max_stages ?

It's not related to the stage, it's the max file numbers in one stage.

**Comment by TCeason** [comment]
> A version and a corresponding test must be added if proto changes: see: src/meta/proto-conv/src/util.rs
> 
> ```
>     (
>         30,
>         "2023-02-21: Add: config.proto/WebhdfsStorageConfig; Modify: user.proto/UserStageInfo::StageStorage",
>     ),
>     // Dear developer:
>     //      If you're gonna add a new metadata version, you'll have to add a test for it.
>     //      You could just copy an existing test file(e.g., `../tests/it/v024_table_meta.rs`)
>     //      and replace two of the variable `bytes` and `want`.
> ```

Done

**Comment by youngsofun** [comment]
better if add a test for it

**Comment by youngsofun** [comment]
better if add a test for it

**Comment by youngsofun** [comment]
better if add a test for it

**Comment by youngsofun** [comment]
https://github.com/datafuselabs/databend/pull/10790/files#diff-dab950962c824e1e8eb5e42ad3ca711b93add5b55758e27bb3168b5ddc9e5b1bR545

https://github.com/datafuselabs/databend/pull/10790/files#diff-de6f99c44be45df8f59ec686a3efc23697c5acb20b14e2da9cee7957346978a0R135

if brach can be remoted?

**Comment by youngsofun** [comment]
https://github.com/datafuselabs/databend/pull/10790/files#diff-dab950962c824e1e8eb5e42ad3ca711b93add5b55758e27bb3168b5ddc9e5b1bR545

https://github.com/datafuselabs/databend/pull/10790/files#diff-de6f99c44be45df8f59ec686a3efc23697c5acb20b14e2da9cee7957346978a0R135

if brach can be removed?

**Comment by youngsofun** [comment]
https://github.com/datafuselabs/databend/pull/10790/files#diff-a1542f638b6f35e3b76275a191596134dc321ac44f226f4b6d54c3dad6dd9920R757

FileLimit -> MaxFiles

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 2]
```diff
@@ -246,7 +246,13 @@ impl CopyInterpreter {
         }
 
         let mut stage_table_info = stage_table_info.clone();
-        let mut all_source_file_infos = StageTable::list_files(&stage_table_info).await?;
+        let max_files = stage_table_info.stage_info.copy_options.max_files;
+
+        let mut all_source_file_infos = if force {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy.rs:251]: Why to check `force` here?

**Comment by BohuTANG** [comment]
Please add a test to cover the max file, and keep it right :)

File: src/query/service/src/interpreters/interpreter_copy.rs (hunk 2 above)
Comment (TCeason) [reply src/query/service/src/interpreters/interpreter_copy.rs:251]: if force = false, will into color_copied_files, limit file in color_copied_files.

if force = true, will not into color_copied_files, limit file in list_files.

Comment (sundy-li) [reply src/query/service/src/interpreters/interpreter_copy.rs:251]: Seems it's no relation with force

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 3]
```diff
@@ -246,7 +246,13 @@ impl CopyInterpreter {
         }
 
         let mut stage_table_info = stage_table_info.clone();
-        let mut all_source_file_infos = StageTable::list_files(&stage_table_info).await?;
+        let max_files = stage_table_info.stage_info.copy_options.max_files;
```
Comment (sundy-li) [inline src/query/service/src/interpreters/interpreter_copy.rs:249]: If max_files is zero, we can set to `usize::max`, the following code do not need to check zero now.

File: src/common/storage/src/stage.rs (hunk 1 above)
Comment (sundy-li) [reply src/common/storage/src/stage.rs]: It's hard to track, I think the users should use `purge = 1` to make it right.  So it's reasonable to use limit here.

File: src/query/service/src/interpreters/interpreter_user_stage_remove.rs [hunk 4]
```diff
@@ -59,7 +59,7 @@ impl Interpreter for RemoveUserStageInterpreter {
             pattern,
         };
         let files: Vec<String> = files_info
-            .list(&op, false)
+            .list(&op, false, 0)
```
Comment (sundy-li) [inline src/query/service/src/interpreters/interpreter_user_stage_remove.rs:62]: usize::max

File: src/query/service/src/table_functions/list_stage/list_stage_table.rs [hunk 5]
```diff
@@ -189,7 +189,7 @@ impl AsyncSource for ListStagesSource {
             pattern: self.args_parsed.files_info.pattern.clone(),
         };
 
-        let files = files_info.list(&op, false).await?;
+        let files = files_info.list(&op, false, 0).await?;
```
Comment (sundy-li) [inline src/query/service/src/table_functions/list_stage/list_stage_table.rs:192]: usize::max

File: src/query/sql/src/planner/binder/copy.rs [hunk 6]
```diff
@@ -542,7 +542,7 @@ impl<'a> Binder {
         let mut files = if operator.info().can_blocking() {
             files_info.blocking_list(&operator, false)
         } else {
-            files_info.list(&operator, false).await
+            files_info.list(&operator, false, 0).await
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/copy.rs:545]: usize::max

File: src/query/storages/parquet/src/parquet_table/partition.rs [hunk 7]
```diff
@@ -115,7 +115,7 @@ impl ParquetTable {
             None => if self.operator.info().can_blocking() {
                 self.files_info.blocking_list(&self.operator, false)
             } else {
-                self.files_info.list(&self.operator, false).await
+                self.files_info.list(&self.operator, false, 0).await
```
Comment (sundy-li) [inline src/query/storages/parquet/src/parquet_table/partition.rs:118]: usize::max

File: src/query/storages/stage/src/stage_table.rs [hunk 8]
```diff
@@ -117,7 +120,7 @@ impl Table for StageTable {
         let files = if let Some(files) = &stage_info.files_to_copy {
             files.clone()
         } else {
-            StageTable::list_files(stage_info).await?
+            StageTable::list_files(stage_info, 0_usize).await?
```
Comment (sundy-li) [inline src/query/storages/stage/src/stage_table.rs:123]: .

File: src/query/service/src/interpreters/interpreter_user_stage_remove.rs (hunk 4 above)
Comment (TCeason) [reply src/query/service/src/interpreters/interpreter_user_stage_remove.rs:62]: Yes, chang to Option<usize> is better.

```rust
pub async fn list_files(
        stage_info: &StageTableInfo,
        max_files: Option<usize>,
    )
```

File: src/common/storage/src/stage.rs [hunk 9]
```diff
@@ -107,9 +107,16 @@ impl StageFilesInfo {
...
+    pub async fn list(
+        &self,
+        operator: &Operator,
+        first_only: bool,
+        max_files: Option<usize>,
+    ) -> Result<Vec<StageFileInfo>> {
+        let max_files = max_files.unwrap_or(usize::MAX);
         if let Some(files) = &self.files {
             let mut res = Vec::new();
+            let mut limit: usize = 0;
```
Comment (youngsofun) [inline src/common/storage/src/stage.rs:119]: not have to change, but a little strange to call it `limit`,   : )

**Review by sundy-li — APPROVED** [review]

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+EOF
+
+cat << EOF > /tmp/force_false.csv
+"id","name","upadte_at","content"
+3,"a","2022-10-25 10:51:14","{\"hello\":\"world\"}"
+4,"b","2022-10-25 10:51:28","{\"金角大王\":\"沙河\"}"
+EOF
+
+
+aws --endpoint-url http://127.0.0.1:9900/ s3 cp /tmp/force_true.csv s3://testbucket/admin/data/csv/force_true.csv > /dev/null 2>&1
```
Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh:32]: no need for aws here

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  c3 VARCHAR
+) ENGINE=FUSE;" | $MYSQL_CLIENT_CONNECT
+
+cat << EOF > /tmp/force_true.csv
+"id","name","upadte_at","content"
+1,"a","2022-10-25 10:51:14","{\"hello\":\"world\"}"
+2,"c","2022-10-25 10:51:50","{\"银角大王\":\"沙河\"}"
+EOF
+
+cat << EOF > /tmp/force_false.csv
```
Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh:25]:
```
rm -rf /tmp/05_02-06
mkdir /tmp/05_02-06
cat << EOF > /tmp/05_02-06/f1.csv
1,2
3,4
EOF
cat << EOF > /tmp/05_02-06/f1.csv
5,6
7,8
EOF

Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh:25]:
```
rm -rf /tmp/05_02-06
mkdir /tmp/05_02-06
cat << EOF > /tmp/05_02-06/f1.csv
1,2
3,4
EOF
cat << EOF > /tmp/05_02-06/f1.csv
5,6
7,8
EOF

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh (hunk 10 above)
Comment (TCeason) [reply tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh]: Fixed.

**Review by youngsofun — APPROVED** [review]

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.result [hunk 12]
```diff
@@ -1,2 +1,3 @@
 4
 4
+4
```
Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.result:3]: expect 6

File: src/query/service/src/sessions/query_ctx.rs [hunk 13]
```diff
@@ -516,9 +518,14 @@ impl TableContext for QueryContext {
             copied_files.extend(resp.file_info);
         }
 
+        let max_files = max_files.unwrap_or(usize::MAX);
         // Colored.
         let mut results = Vec::with_capacity(files.len());
         for mut file in files {
+            if limit == max_files {
```
Comment (youngsofun) [inline src/query/service/src/sessions/query_ctx.rs:525]: @TCeason @sundy-li  bug here

**Pull request merged**